### DISCUSSION TITLE
Integrate NDISAPI, An Alternative for Tun (Windows)

### DISCUSSION THREAD

**Discussion by amir-devman** [XTLS/Xray-core#4228] [Ideas]
I would like to propose the integration of NDISAPI into Xray-core as an alternative to the TUN driver for Windows users.

**Why [NDISAPI](https://github.com/wiresock/ndisapi-go)?**
NDISAPI is a high-performance packet filtering framework for Windows. Benchmarks on [WireSock.net](https://wiresock.net/) demonstrate that NDISAPI outperforms traditional TUN implementations, including WireGuard TUN, in terms of throughput and CPU efficiency. Integrating NDISAPI into Xray-core could provide users with a significant performance boost, especially for high-speed or resource-constrained scenarios.

**Benefits of NDISAPI Integration:**
Superior Performance: Lower CPU utilization and higher throughput compared to TUN.
Advanced Features: Direct control of packet routing, filtering, and modification at the driver level, enabling more efficient traffic management.
Optimized for Windows: NDISAPI is specifically designed for Windows, making it a natural choice for enhancing the performance and flexibility of Xray-core on this platform.
Proposal Details
I have already developed a Go implementation of NDISAPI, which is ready to integrate into projects like Xray-core. I am also willing to personally implement and maintain this feature through a PR.

**Community Feedback**
Before proceeding, I would like to gather feedback from the maintainers and the community:

**Would this feature align with Xray-core’s goals and priorities?**
Are there any specific concerns or prerequisites to consider for implementing this feature?
I believe this integration could greatly benefit Windows users of Xray-core by providing a powerful and efficient alternative to TUN. I look forward to hearing your thoughts and engaging in a constructive discussion!

Thank you for your time and consideration.

**Comment by Smallthing** [comment]
I want to try it

**Comment by RPRX** [comment]
PR is welcomed

**Reply by amir-devman** [reply]
I have a quick question, should I add it as `feature` or `inbound`?

**Reply by RPRX** [reply]
`inbound`

**Reply by amir-devman** [reply]
Another question too, there are two ways to implement this. Having a transparent proxy to redirect data to it or use gvisor stack to process packets:

GVisor example written by nekohasekai
https://github.com/SagerNet/sing-box/tree/dev-ndis

And redirection by me:
https://github.com/amir-devman/sing-box/commits/dev-next

Both can work perfectly

**Reply by RPRX** [reply]
我没理解错的话这两种方式都被 Windows Packet Filter Driver 所支持，不过我刚刚看了下这个 Driver 严格来说不是免费的，我觉得自己弄个虚拟网卡再用 gVisor 来处理流量并不难

**Reply by RPRX** [reply]
不过我还没研究怎么做到区分进程的，我觉得仅虚拟网卡似乎不能区分进程？

**Reply by amir-devman** [reply]
Do you mean, how to bypass xray-core traffic to not to go through NDIS?

**Reply by RPRX** [reply]
@amir-devman 差不多，我猜应该有某种机制在传递文件路径

**Reply by amir-devman** [reply]
Yes, I've already implemented the tools to bypass traffic of certain processes (which this can allow you to bypass itself process - auto route) using path or .exe filename. That's done by doing some `syscalls`.

https://github.com/wiresock/ndisapi-go/tree/main/netlib

**Comment by romanholidaypancakes** [comment]
ping @amir-devman  https://github.com/XTLS/Xray-core/issues/4292

**Reply by amir-devman** [reply]
I've been so busy with my current works, I don't think I can do it anytime soon.
